### PR TITLE
feature(eos_designs): Add routed interfaces to VRF

### LINKED ISSUE #836: eos_designs: l3edge: add variable for vrf
**Issue Type**

- [x] Template enhancement
- [ ] Role enhancement
- [ ] Documentation enhancement (avd.sh)

**Describe the solution you'd like**

eos_designs/l3edge template allows to configure L3 P2P links anywhere in the fabric to any device.
We would need to be able to move such a L3 P2P interface into a certain VRF.
The result should look like this:

```yaml
  Ethernet5:
    peer: core-01
    peer_interface: gi1
    peer_type: other
    description: P2P_LINK_TO_core-01_gi1
    type: routed
    shutdown: false
    mtu: 9000
    ip_address: 10.130.8.46/31
    vrf: Tenant_DC_OUTSIDE
 ```

**Describe alternatives you've considered**

I could set it via customer config in host_vars files, but with this the config is at different places and might lead
to missunderstanding or even loss of information for the users.

custom_structured_configuration_ethernet_interfaces:

```yaml
  Ethernet5:
    vrf: Tenant_DC_OUTSIDE
```

**Additional context**

This might already do what I am watching for:

\avd\roles\eos_designs\templates\l3_edge\switch-l3-edge-p2p-interfaces.j2
{%     if p2p_interface.vrf is arista.avd.defined %}
    vrf: {{ p2p_interface.vrf }}
{%     endif %}

\avd\roles\eos_designs\templates\l3_edge\logic\switch-l3-edge.j2
{%             set p2p_interface.vrf = p2p_link.vrf | arista.avd.default(
                                              p2p_profile.vrf) %}

### PR DESCRIPTION
**Pull request by mpergament**
## Change Summary

Add routed interfaces to VRF

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Documentation content changes
- [ ] Other (please describe):

## Related Issue(s)

Fixes #836 

## Component(s) name

`arista.avd.eos_designs`

## Proposed changes
```
    Tenant_A_WAN_Zone:
        vrf_vni: 14
        l3_interfaces:
          - interfaces: [ Ethernet7, Ethernet7 ]
            ip_addresses: [10.10.10.10/24, 10.10.20.20/24]
            nodes: [DC1-BL1A, DC1-BL1B]
            mtu: 9000
            description: "test"
```
## Checklist:

- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://www.avd.sh/docs/contributing/) document.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have updated [molecule CI](https://github.com/aristanetworks/ansible-avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly
- [x] All new and existing tests passed ([`pre-commit`](https://www.avd.sh/docs/installation/development/#python-virtual-environment), `make linting` and `make sanity-lint`).

### CODE REVIEW

File: ansible_collections/arista/avd/molecule/evpn_underlay_ebgp_overlay_ebgp/documentation/devices/DC1-BL1A.md [hunk 1]
```diff
@@ -353,6 +354,13 @@ interface Ethernet4
    no switchport
    ip address 172.31.255.87/31
 !
+interface Ethernet7
+   description test
+   mtu 9000
+   no switchport
+   vrf Tenant_A_WAN_Zone
+   ip address 10.10.10.10/24
+!
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/molecule/evpn_underlay_ebgp_overlay_ebgp/documentation/devices/DC1-BL1A.md:366]: Missing no shutdown in generated config. I think you need to add `enabled: true`

**Review by ClausHolbechArista — COMMENTED** [review]
One comment so far. I will review more thoroughly tonight.

File: ansible_collections/arista/avd/roles/eos_designs/templates/designs/l3ls-evpn/tenant_evpn_vxlan/l3leaf-logic.j2 [hunk 2]
```diff
@@ -59,6 +59,25 @@
...
+{% for tenant in tenants | arista.avd.natural_sort if tenant in leaf.filter_tenants or "all" in leaf.filter_tenants %}
+{%     if tenants[tenant].vrfs is arista.avd.defined %}
+{%         for vrf in tenants[tenant].vrfs if vrf in leaf.vrfs %}
+{%             if tenants[tenant].vrfs[vrf].l3_interfaces is arista.avd.defined %}
+{%                 for entry in tenants[tenant].vrfs[vrf].l3_interfaces %}
+{%                     if entry.nodes is arista.avd.defined and entry.ip_addresses is arista.avd.defined and entry.interfaces is arista.avd.defined and inventory_hostname in entry.nodes %}
+{%                         set node_index = entry.nodes.index(inventory_hostname) %}
+{%                         do entry.update( {'vrf':vrf} )%}
+{%                         do entry.update( {'l3_interface':entry.interfaces[node_index]} )%}
+{%                         do entry.update( {'ip_address':entry.ip_addresses[node_index]})%}
```
Comment (emilotep) [inline ansible_collections/arista/avd/roles/eos_designs/templates/designs/l3ls-evpn/tenant_evpn_vxlan/l3leaf-logic.j2]: Really minor but some inconsistent spacing after closing "}" and between closing ")" and "%}".

Comment (emilotep) [reply ansible_collections/arista/avd/roles/eos_designs/templates/designs/l3ls-evpn/tenant_evpn_vxlan/l3leaf-logic.j2]: Approved, LGTM.

**Review by ankudinov — APPROVED** [review]
LGTM

**Comment by daanvdsanden** [comment]
Does not seem to work is there are only l3 interfaces in the VRF on a switch. Either always_include_vrfs_in_tenants has to be defined for that switch or an SVI also needs to be present in the VRF else the vrf will not be in leaf.vrfs:
https://github.com/aristanetworks/ansible-avd/blob/7ce00e648d3ee8b7dc28839650874d5927710b82/ansible_collections/arista/avd/roles/eos_designs/templates/designs/l3ls-evpn/tenant_evpn_vxlan/l3leaf-logic.j2#L66

If below code block is moved to line 18 and a line is added to add the vrf to the leaf.vrfs array, it behaves more like an svi (and always_include_vrfs_in_tenants doesn't need to be set)
https://github.com/aristanetworks/ansible-avd/blob/7ce00e648d3ee8b7dc28839650874d5927710b82/ansible_collections/arista/avd/roles/eos_designs/templates/designs/l3ls-evpn/tenant_evpn_vxlan/l3leaf-logic.j2#L67-L77

**Comment by mpergament** [comment]
> Does not seem to work is there are only l3 interfaces in the VRF on a switch. Either always_include_vrfs_in_tenants has to be defined for that switch or an SVI also needs to be present in the VRF else the vrf will not be in leaf.vrfs:
> https://github.com/aristanetworks/ansible-avd/blob/7ce00e648d3ee8b7dc28839650874d5927710b82/ansible_collections/arista/avd/roles/eos_designs/templates/designs/l3ls-evpn/tenant_evpn_vxlan/l3leaf-logic.j2#L66
> 
> If below code block is moved to line 18 and a line is added to add the vrf to the leaf.vrfs array, it behaves more like an svi (and always_include_vrfs_in_tenants doesn't need to be set)
> https://github.com/aristanetworks/ansible-avd/blob/7ce00e648d3ee8b7dc28839650874d5927710b82/ansible_collections/arista/avd/roles/eos_designs/templates/designs/l3ls-evpn/tenant_evpn_vxlan/l3leaf-logic.j2#L67-L77

Good catch!!! Thanks a lot for your review!

**Comment by daanvdsanden** [comment]
When the node names are not unique, only the first interface on that node is used. Not sure if you want to support below configuration (might mention it in the documentation). 

e.g. the following will only configure interface Ethernet7 and not Ethernet8
```
    Tenant_A_WAN_Zone:
        vrf_vni: 14
        l3_interfaces:
          - interfaces: [ Ethernet7, Ethernet8 ]
            ip_addresses: [10.10.10.10/24, 10.10.20.20/24]
            nodes: [DC1-BL1A, DC1-BL1A]
            mtu: 9000
            description: "test"
```

**Comment by mpergament** [comment]
> When the node names are not unique, only the first interface on that node is used. Not sure if you want to support below configuration (might mention it in the documentation).
> 
> e.g. the following will only configure interface Ethernet7 and not Ethernet8
> 
> ```
>     Tenant_A_WAN_Zone:
>         vrf_vni: 14
>         l3_interfaces:
>           - interfaces: [ Ethernet7, Ethernet8 ]
>             ip_addresses: [10.10.10.10/24, 10.10.20.20/24]
>             nodes: [DC1-BL1A, DC1-BL1A]
>             mtu: 9000
>             description: "test"
> ```

Let me add note to the documentation

**Comment by carlbuchmann** [comment]
> > When the node names are not unique, only the first interface on that node is used. Not sure if you want to support below configuration (might mention it in the documentation).
> > e.g. the following will only configure interface Ethernet7 and not Ethernet8
> > ```
> >     Tenant_A_WAN_Zone:
> >         vrf_vni: 14
> >         l3_interfaces:
> >           - interfaces: [ Ethernet7, Ethernet8 ]
> >             ip_addresses: [10.10.10.10/24, 10.10.20.20/24]
> >             nodes: [DC1-BL1A, DC1-BL1A]
> >             mtu: 9000
> >             description: "test"
> > ```
> 
> Let me add note to the documentation

Please update the j2 template to handle this scenario, so it's consistent with similar data models

**Review by carlbuchmann — DISMISSED** [review]
See comment in the "Conversation" board!

**Comment by mpergament** [comment]
> See comment in the "Conversation" board!

Done

**Review by titom73 — APPROVED** [review]
LGTM!

File: ansible_collections/arista/avd/molecule/evpn_underlay_ebgp_overlay_ebgp/documentation/fabric/DC1_FABRIC-topology.csv [hunk 3]
```diff
@@ -3,11 +3,17 @@ l3leaf,DC1-BL1A,Ethernet1,spine,DC1-SPINE1,Ethernet6
 l3leaf,DC1-BL1A,Ethernet2,spine,DC1-SPINE2,Ethernet6
 l3leaf,DC1-BL1A,Ethernet3,spine,DC1-SPINE3,Ethernet6
 l3leaf,DC1-BL1A,Ethernet4,spine,DC1-SPINE4,Ethernet6
+l3leaf,DC1-BL1A,Ethernet7,,,
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/molecule/evpn_underlay_ebgp_overlay_ebgp/documentation/fabric/DC1_FABRIC-topology.csv]: These CSVs reflect the topology. We should at least add some static "peer_type" key on the generated structured config. Ex. "l3_interface"

File: ansible_collections/arista/avd/molecule/evpn_underlay_ebgp_overlay_ebgp/inventory/group_vars/DC1_TENANTS_NETWORKS.yml [hunk 4]
```diff
@@ -96,6 +96,13 @@ tenants:
             ip_address_virtual: 10.1.41.1/24
       Tenant_A_WAN_Zone:
         vrf_vni: 14
+        l3_interfaces:
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/molecule/evpn_underlay_ebgp_overlay_ebgp/inventory/group_vars/DC1_TENANTS_NETWORKS.yml:99]: to cover more scenarios, could we change this section to be one list item per interface, so we also test the outer loop over l3_interfaces ?

File: ansible_collections/arista/avd/roles/eos_designs/templates/designs/l3ls-evpn/tenant_evpn_vxlan/l3leaf-logic.j2 [hunk 5]
```diff
@@ -59,6 +59,31 @@
...
+{%     if tenants[tenant].vrfs is arista.avd.defined %}
+{%         for vrf in tenants[tenant].vrfs %}
+{%             if tenants[tenant].vrfs[vrf].l3_interfaces is arista.avd.defined %}
+{%                 for entry in tenants[tenant].vrfs[vrf].l3_interfaces %}
+{%                     if entry.nodes is arista.avd.defined and entry.ip_addresses is arista.avd.defined and entry.interfaces is arista.avd.defined and inventory_hostname in entry.nodes %}
+{%                         for node in entry.nodes %}
+{%                             if node == inventory_hostname %}
+{%                                 set entry_node = dict() %}
+{%                                 do entry_node.update( {'vrf':vrf} )%}
+{%                                 do entry_node.update( {'l3_interface':entry.interfaces[loop.index - 1]} ) %}
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/templates/designs/l3ls-evpn/tenant_evpn_vxlan/l3leaf-logic.j2]: use loop.index0 to simplify.

File: ansible_collections/arista/avd/roles/eos_designs/templates/designs/l3ls-evpn/tenant_evpn_vxlan/leaf-tenant-vrf-interfaces.j2 [hunk 6]
```diff
@@ -0,0 +1,17 @@
+{% if leaf.l3_interfaces is defined %}
+{%     for interface in leaf.l3_interfaces %}
+  {{ interface.l3_interface }}:
+    type: routed
+    vrf: {{ interface.vrf }}
+    ip_address: {{ interface.ip_address }}
+{%         if interface.mtu is arista.avd.defined %}
+    mtu: {{ interface.mtu }}
+{%         endif %}
+{%         if interface.enabled is arista.avd.defined(true) %}
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/templates/designs/l3ls-evpn/tenant_evpn_vxlan/leaf-tenant-vrf-interfaces.j2]: I suppose the default should be enabled? If so we should test for `false` and set `shutdown: true` and the `else -> shutdown: false`. Please don't rely on EOS defaults, as things like shutdown can be easily reversed with global command.

**Review by ClausHolbechArista — CHANGES_REQUESTED** [review]
A few changes requested.

**Review by ClausHolbechArista — APPROVED** [review]
LGTM

**Comment by kmueller68** [comment]
Thank you very much to all of you for your effort implementing this.
This helps me very much to setup things more effective for my current project.

**Comment by gusmb** [comment]
> > > When the node names are not unique, only the first interface on that node is used. Not sure if you want to support below configuration (might mention it in the documentation).
> > > e.g. the following will only configure interface Ethernet7 and not Ethernet8
> > > ```
> > >     Tenant_A_WAN_Zone:
> > >         vrf_vni: 14
> > >         l3_interfaces:
> > >           - interfaces: [ Ethernet7, Ethernet8 ]
> > >             ip_addresses: [10.10.10.10/24, 10.10.20.20/24]
> > >             nodes: [DC1-BL1A, DC1-BL1A]
> > >             mtu: 9000
> > >             description: "test"
> > > ```
> > 
> > 
> > Let me add note to the documentation
> 
> Please update the j2 template to handle this scenario, so it's consistent with similar data models

How did this end up being resolved? I would expect to manage "non-uniqueness" this way:

 ```
Tenant_A_WAN_Zone:
   vrf_vni: 14
   l3_interfaces:
     - interfaces: [ Ethernet7, Ethernet8 ]
        ip_addresses: [10.10.10.10/24, 10.10.20.20/24]
        nodes: [DC1-BL1A, DC1-BL2A]
        mtu: 9000
        description: "test"
     - interfaces: [ Ethernet9 ]
       ip_addresses: [10.10.30.30/24]
       nodes: [DC1-BL1A]
       mtu: 9000
       description: "test"
```

Since `l3_interfaces` is a list, it should support multiple list elements with unique keys. Like in the example above, Ethernet9 l3 interface is configured on the same node as Ethernet7. In general for this case, wouldn't it be simpler to have one interface per list item? This would probably provide more clarity. Example:

 ```
Tenant_A_WAN_Zone:
   vrf_vni: 14
   l3_interfaces:
     - interface: Ethernet7
        ip_address: 10.10.10.10/24
        node: DC1-BL1A
        mtu: 9000
        description: "test"
     - interface: Ethernet8
        ip_address: 10.10.20.20/24
        node: DC1-BL2A
        mtu: 9000
        description: "test"
     - interface: Ethernet9
        ip_address: 10.10.30.30/24
        node: DC1-BL1A
        mtu: 9000
        description: "test"
```

Then there is no conflict about the uniqueness, as every list element can be uniquely identified by it's key (ip_address, which is unique per VRF)

**Pull request merged**